lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -2227,6 +2232,50 @@ balanceTransactionSpec = do
             it "may fail"
                 $ property prop_balanceTransactionUnresolvedInputs
 
+    describe "sizeOfCoin" $ do
+        let coinToWord64Clamped = fromMaybe maxBound . Coin.toWord64
+        let cborSizeOfCoin =
+                TxSize
+                . fromIntegral
+                . BS.length
+                . CBOR.toStrictByteString
+                . CBOR.encodeWord64 . coinToWord64Clamped
+
+        let isBoundary c =
+                sizeOfCoin c /= sizeOfCoin (c `Coin.difference` (Coin 1))
+                || sizeOfCoin c /= sizeOfCoin (c `Coin.add` (Coin 1))
+
+        it "matches the size of the Word64 CBOR encoding" $
+            property $ checkCoverage $
+                forAll genEncodingBoundaryLovelace $ \l -> do
+                    let c = fromCardanoLovelace l
+                    let expected = cborSizeOfCoin c
+
+                    -- Use a low coverage requirement of 0.01% just to
+                    -- ensure we see /some/ amount of every size.
+                    let coverSize s = cover 0.01 (s == expected) (show s)
+                    sizeOfCoin c === expected
+                        & coverSize (TxSize 1)
+                        & coverSize (TxSize 2)
+                        & coverSize (TxSize 3)
+                        & coverSize (TxSize 5)
+                        & coverSize (TxSize 9)
+                        & cover 0.5 (isBoundary c) "boundary case"
+
+        describe "boundary case goldens" $ do
+            it "1 byte to 2 byte boundary" $ do
+                sizeOfCoin (Coin 23) `shouldBe` TxSize 1
+                sizeOfCoin (Coin 24) `shouldBe` TxSize 2
+            it "2 byte to 3 byte boundary" $ do
+                sizeOfCoin (Coin 255) `shouldBe` TxSize 2
+                sizeOfCoin (Coin 256) `shouldBe` TxSize 3
+            it "3 byte to 5 byte boundary" $ do
+                sizeOfCoin (Coin 65535) `shouldBe` TxSize 3
+                sizeOfCoin (Coin 65536) `shouldBe` TxSize 5
+            it "5 byte to 9 byte boundary" $ do
+                sizeOfCoin (Coin 4294967295) `shouldBe` TxSize 5
+                sizeOfCoin (Coin 4294967296) `shouldBe` TxSize 9
```

Comment:
Perhaps explicitly indicate that these are powers of 2?

Justification: I suspect it better expresses the intent, and it also emphasises the difference between each pair (which is just `1`), which is harder to see visually if we write out the whole number and need to compare every digit.

```suggestion
            it "2 byte to 3 byte boundary" $ do
                sizeOfCoin (Coin $ 2^8 - 1) `shouldBe` TxSize 2
                sizeOfCoin (Coin $ 2^8    ) `shouldBe` TxSize 3
            it "3 byte to 5 byte boundary" $ do
                sizeOfCoin (Coin $ 2^16 - 1) `shouldBe` TxSize 3
                sizeOfCoin (Coin $ 2^16    ) `shouldBe` TxSize 5
            it "5 byte to 9 byte boundary" $ do
                sizeOfCoin (Coin $ 2^32 - 1) `shouldBe` TxSize 5
                sizeOfCoin (Coin $ 2^32    ) `shouldBe` TxSize 9
```